x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java

```diff
@@ -241,11 +246,11 @@ public Authentication maybeRewriteForOlderVersion(Version olderVersion) {
      */
     public Authentication runAs(User runAs, @Nullable RealmRef lookupRealmRef) {
         Objects.requireNonNull(runAs);
-        assert false == runAs.isRunAs();
-        assert false == getUser().isRunAs();
+        assert false == runAs instanceof RunAsUser;
+        assert false == getUser() instanceof RunAsUser;
         assert AuthenticationType.REALM == getAuthenticationType() || AuthenticationType.API_KEY == getAuthenticationType();
```

Comment:
Should we also check that `runAs.getAuthenticationType() == AuthenticationType.REALM` ?

Reply:
Not sure if I miss something, but `runAs` is a `User`, not an `Authentication`. So it does not have an `AuthenticationType`. Do you mean we should check `lookUpRealmRef` is not one of the sythetic realms? Unfortunately we cannot because of custom realms and we don't retrict their names or types.

Reply:
Sorry, yeah - too many classes with things in the wrong place.

Logically we ought to check that `runAs` is a realm based used, and not an Internal user or an API Key, or ServiceAccount.

Since the proposal is an `assert`, I don't think we need to worry about custom realms. ES doesn't run with assertions in production, and if it interferes with someone's development environment then it just serves as a reminder that they picked a terrible name & type for their realm.

Reply:
I added assertion for lookupRealmRef not having synthetic realm name or type. 

PS: I also plan to consolidate all the assertions from different Authentication methods into its own private method and call it inside the constructors. This method will also serve a good reference for the internal logic of authentication object.